Legacy/BootPlatform/DxeIpl/Ia32/Canary.nasm

```diff
@@ -0,0 +1,93 @@
+; @file
+; Copyright (C) 2021, ISP RAS. All rights reserved.
+;
+; This program and the accompanying materials
+; are licensed and made available under the terms and conditions of the BSD License
+; which accompanies this distribution.  The full text of the license may be found at
+; http://opensource.org/licenses/bsd-license.php
+;
+; THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,
+; WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+;
+BITS 32
+
+#include <AutoGen.h>
+extern ASM_PFX(_ModuleEntryPointReal)
+
+section .data
+align 8
+global ASM_PFX(__security_cookie)
+ASM_PFX(__security_cookie):
+  dd 0
+
+global ASM_PFX(__stack_chk_guard)
+ASM_PFX(__stack_chk_guard):
+  dd 0
+
+section .text
+;  BIT4 - Enable BreakPoint as ASSERT. (MdePkg.dec)
+%define DEBUG_PROPERTY_ASSERT_BREAKPOINT_ENABLED ((FixedPcdGet8(PcdDebugPropertyMask) & 0x10) != 0)
+
+; #######################################################################
+; VOID __stack_chk_fail (VOID)
+; #######################################################################
+align 8
+global ASM_PFX(__stack_chk_fail)
+ASM_PFX(__stack_chk_fail):
+%if DEBUG_PROPERTY_ASSERT_BREAKPOINT_ENABLED
+  int 3
+  ret
+%else
+back:
+  cli
+  hlt
+  jmp back
+%endif
+
+; #######################################################################
+; VOID
+; __security_check_cookie (
+;   IN UINTN  Value
+;   )
+; #######################################################################
+align 8
+global ASM_PFX(__security_check_cookie)
+ASM_PFX(__security_check_cookie):
+  mov  ecx, [esp+4] ; get UINTN Value
+  mov  eax, dword [rel ASM_PFX(__security_cookie)]
+  test eax, ecx
+  jnz  ASM_PFX(__stack_chk_fail)
+  ret
+
+; #######################################################################
+; EFI_STATUS
+; EFIAPI
+; _ModuleEntryPoint (
+;   IN EFILDRHANDOFF  *Handoff
+;   )
+; #######################################################################
+align 8
+global ASM_PFX(_ModuleEntryPoint)
+ASM_PFX(_ModuleEntryPoint):
+%if FixedPcdGet8(PcdCanaryAllowRdtscFallback)
+  mov eax, 1          ; Feature Information
+  cpuid               ; result in EAX, EBX, ECX, EDX
+  and ecx, 040000000H
+  cmp ecx, 040000000H ; check RDRAND feature flag
+  jne noRdRand
+retry:
```

Comment:
Use local branches starting from `.` as it may otherwise cause unexpected conflicts.